File: scripts/core/immutability.py
import os
import sys
import subprocess

_depth = 0
_mount_point = "/mnt"


def log(msg):
    print("  " * _depth + msg)


def warn(msg):
    print("  " * _depth + f"WRN {msg}", file=sys.stderr)


def error(msg):
    print("  " * _depth + f"ERR {msg}", file=sys.stderr)


def run(cmd):
    global _depth
    _depth += 1
    log(cmd)
    result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
    for line in result.stdout.splitlines():
        if line.strip():
            log(line)
    if result.returncode != 0:
        for line in result.stderr.splitlines():
            if line.strip():
                warn(line)
    _depth -= 1
    return result.returncode


def abort(msg):
    error(msg)
    error("Unmounting and quitting")
    run(f"umount -R {_mount_point}")
    run(f"rm -rf {_mount_point}")
    sys.exit(1)


def require(test_args):
    if run(f"test {test_args}") != 0:
        abort(f"Require failed: test {test_args}")


def btrfs_sync(path):
    run(f"btrfs filesystem sync {path}")


def btrfs_delete(path):
    if run(f"btrfs subvolume delete {path} --commit-after") != 0:
        abort(f"Failed to delete {path}")
    btrfs_sync(os.path.dirname(path))


def btrfs_delete_recursively(path):
    if not os.path.isdir(path):
        return
    result = subprocess.run(
        f"btrfs subvolume list -o {path}",
        shell=True, capture_output=True, text=True,
    )
    for line in result.stdout.splitlines():
        parts = line.split()
        if len(parts) >= 9:
            subvol = os.path.join(
                os.path.dirname(path), " ".join(parts[8:])
            )
            btrfs_delete_recursively(subvol)
    btrfs_delete(path)


def btrfs_snapshot(src, dst):
    require(f"-d {src}")
    btrfs_delete_recursively(dst)
    if run(f"btrfs subvolume snapshot {src} {dst}") != 0:
        abort(f"Failed to snapshot {src} to {dst}")
    btrfs_sync(src)
    btrfs_sync(dst)


def btrfs_set_rw(path):
    if run(f"btrfs property set -ts {path} ro false") != 0:
        abort(f"Failed to make {path} read-write")


def mount_subvolumes(disk):
    require(f"-b {disk}")
    run(f"mkdir -p {_mount_point}")
    run(f"mount -t btrfs -o subvolid=5,user_subvol_rm_allowed"
        f" {disk} {_mount_point}")


def unmount_subvolumes():
    run(f"umount {_mount_point}")
    run(f"rm -rf {_mount_point}")


def check_recovery_needed(snapshot):
    return (os.path.isdir(snapshot)
            and not os.path.isfile(f"{snapshot}/.boot-ready"))


def create_sentinel(snapshot):
    with open(f"{snapshot}/.boot-ready", "w"):
        pass
    btrfs_sync(snapshot)


def read_paths_file(paths_file):
    with open(paths_file, "r") as f:
        return [line.strip() for line in f if line.strip()]


def build_rsync_filter(mount_point, paths, previous):
    lines = ["+ */"]
    for path in paths:
        if mount_point == "/":
            relative = path.lstrip("/")
        elif path == mount_point or path.startswith(mount_point + "/"):
            relative = path[len(mount_point):].lstrip("/")
        else:
            continue
        if not relative:
            continue
        full = os.path.join(previous, relative)
        if os.path.isdir(full):
            lines.append(f"+ /{relative}/")
            lines.append(f"+ /{relative}/**")
        elif os.path.exists(full):
            lines.append(f"+ /{relative}")
        else:
            log(f"  skip (not in previous): {path}")
    lines.append("- *")
    return lines


def copy_persistent_files(mount_point, paths, previous, current):
    log(f"Preserving persistent paths for mount_point={mount_point}")
    filter_lines = build_rsync_filter(mount_point, paths, previous)
    has_includes = any(
        l.startswith("+ /") and l != "+ */" for l in filter_lines
    )
    if not has_includes:
        log(f"No paths matched mount_point={mount_point}; skipping rsync")
        return
    for line in filter_lines:
        log(f"  filter: {line}")
    filter_path = f"/tmp/immutability-filter-{os.getpid()}"
    with open(filter_path, "w") as f:
        f.write("\n".join(filter_lines) + "\n")
    if run(f"cd {previous} && rsync -aHAX --numeric-ids --delete"
           f" --filter='. {filter_path}' . {current}") != 0:
        abort("rsync failed to copy persistent files")


def reset_subvolume(name, mount_point, snapshots_name, clean_name, paths):
    subvolume = f"{_mount_point}/{name}"
    snapshots = f"{_mount_point}/{snapshots_name}/{name}"
    clean = f"{snapshots}/{clean_name}"
    previous = f"{snapshots}/PREVIOUS"
    penultimate = f"{snapshots}/PENULTIMATE"
    current = f"{snapshots}/CURRENT"

    if check_recovery_needed(current):
        warn("Incomplete boot detected (missing .boot-ready sentinel)")
        btrfs_delete_recursively(current)

    require(f"-n {clean}")
    require(f"-d {clean}")
    if not os.path.isdir(penultimate):
        btrfs_snapshot(clean, penultimate)
    if not os.path.isdir(previous):
        btrfs_snapshot(clean, previous)

    btrfs_snapshot(previous, penultimate)
    btrfs_snapshot(subvolume, previous)
    btrfs_snapshot(clean, current)
    btrfs_set_rw(current)
    copy_persistent_files(mount_point, paths, previous, current)
    create_sentinel(current)
    btrfs_snapshot(current, subvolume)


def restore_subvolume(name, snapshots_name, snapshot_label):
    subvolume = f"{_mount_point}/{name}"
    snapshots = f"{_mount_point}/{snapshots_name}/{name}"
    source = f"{snapshots}/{snapshot_label}"

    if not os.path.isdir(source):
        abort(f"Cannot restore: {source} does not exist")

    log(f"Restoring {name} from {snapshot_label}")
    btrfs_snapshot(source, subvolume)
    log(f"Restored {name} from {snapshot_label}")


def snapshot_only(name, snapshots_name, clean_name):
    subvolume = f"{_mount_point}/{name}"
    snapshots = f"{_mount_point}/{snapshots_name}/{name}"
    clean = f"{snapshots}/{clean_name}"
    previous = f"{snapshots}/PREVIOUS"
    penultimate = f"{snapshots}/PENULTIMATE"

    require(f"-n {clean}")
    require(f"-d {clean}")
    if not os.path.isdir(penultimate):
        btrfs_snapshot(clean, penultimate)
    if not os.path.isdir(previous):
        btrfs_snapshot(clean, previous)

    btrfs_snapshot(previous, penultimate)
    btrfs_snapshot(subvolume, previous)
    log(f"Snapshot-only complete for {name} (no wipe)")


def main():
    if len(sys.argv) < 6:
        error(f"Usage: {sys.argv[0]} <device> <snapshots_name>"
              " <clean_name> <mode> <paths_file> [name=mount ...]")
        sys.exit(1)

    device = sys.argv[1]
    snapshots_name = sys.argv[2]
    clean_name = sys.argv[3]
    mode = sys.argv[4]
    paths_file = sys.argv[5]
    pairs = sys.argv[6:]
    subvolume_names = [p.split("=")[0] for p in pairs]

    log(f"Mode={mode} device={device}"
        f" subvolumes={' '.join(subvolume_names)}")
    mount_subvolumes(device)

    if mode == "disabled":
        log("Immutability disabled; skipping all operations")
        unmount_subvolumes()
        return

    paths = read_paths_file(paths_file)

    for pair in pairs:
        parts = pair.split("=", 1)
        name = parts[0]
        mount_point = parts[1] if len(parts) > 1 else "/"

        if mode == "reset":
            log(f"Resetting {name} (mount_point={mount_point})")
            reset_subvolume(
                name, mount_point, snapshots_name, clean_name, paths,
            )
        elif mode == "snapshot-only":
            log(f"Snapshot-only {name}")
            snapshot_only(name, snapshots_name, clean_name)
        elif mode in ("restore-previous", "restore-penultimate"):
            label = mode.split("-", 1)[1].upper()
            log(f"Restoring {name} from {label}")
            restore_subvolume(name, snapshots_name, label)
        else:
            abort(f"Unknown mode: {mode}")

    unmount_subvolumes()
    log("Immutability complete")


if __name__ == "__main__":
    main()

File: scripts/lib/immutability.rs
// BTRFS subvolume reset for NixOS immutable boot
// Usage: immutability <device> <snap_name> <clean_name> <mode> [name=mp:filter ...]
//
// Key change from original: reset() now snapshots CLEAN→CURRENT (not PREVIOUS→CURRENT),
// then overlays only the persistent paths from PREVIOUS. This avoids walking the entire
// volume tree and instead touches only O(persistent_paths) entries.
use std::{env, fs, path::Path, process::{self, Command}, thread, time::Instant};

const MOUNT_PATH: &str = "/mnt";

fn elapsed(start: &Instant) -> String { format!("[{:.2}s]", start.elapsed().as_secs_f64()) }

fn run(start: &Instant, args: &[&str]) -> bool {
	println!("{} {}", elapsed(start), args.join(" "));
	let output = Command::new(args[0]).args(&args[1..]).output().expect("exec failed");
	for line in String::from_utf8_lossy(&output.stdout).lines().filter(|l| !l.is_empty()) {
		println!("{}   {line}", elapsed(start));
	}
	for line in String::from_utf8_lossy(&output.stderr).lines().filter(|l| !l.is_empty()) {
		eprintln!("{}   {line}", elapsed(start));
	}
	output.status.success()
}

// Panics rather than process::exit so that thread::scope can wait for sibling threads
// to finish their operations before the panic propagates to main. The Unmounter RAII
// guard in main then runs umount on the way out, regardless of the exit path.
fn die(start: &Instant, message: &str) -> ! {
	eprintln!("{} ERR {message}", elapsed(start));
	panic!("immutability fatal: {message}");
}

fn require(start: &Instant, args: &[&str]) {
	if !run(start, args) { die(start, &format!("Failed: {}", args.join(" "))); }
}

// Umounts MOUNT_PATH on drop, covering both normal exit and die() panic paths.
struct Unmounter<'a>(&'a Instant);
impl Drop for Unmounter<'_> {
	fn drop(&mut self) { run(self.0, &["umount", "-R", MOUNT_PATH]); }
}

fn delete_subvolume(start: &Instant, path: &str) {
	if !Path::new(path).is_dir() { return; }
	let output = Command::new("btrfs").args(["subvolume", "list", "-o", path]).output().unwrap();
	for line in String::from_utf8_lossy(&output.stdout).lines() {
		if let Some(child_path) = line.split(" path ").nth(1) {
			delete_subvolume(start, &format!("{MOUNT_PATH}/{child_path}"));
		}
	}
	require(start, &["btrfs", "subvolume", "delete", path]);
}

fn create_snapshot(start: &Instant, source: &str, destination: &str) {
	if !Path::new(source).is_dir() { die(start, &format!("Source missing: {source}")); }
	delete_subvolume(start, destination);
	require(start, &["btrfs", "subvolume", "snapshot", source, destination]);
}

fn parse_persistent_paths(start: &Instant, filter_path: &str) -> Vec<String> {
	if filter_path.is_empty() { return vec![]; }
	let content = match fs::read_to_string(filter_path) {
		Ok(c)  => c,
		Err(e) => die(start, &format!("Cannot read filter {filter_path}: {e}")),
	};
	content.lines()
		.filter(|line| line.starts_with("+ /"))
		.map(|line| line[2..].trim().to_string())
		.filter(|path| !path.ends_with('/') && !path.ends_with("/**"))
		.collect()
}

// Copies each persistent path from previous_root into current_root, replacing whatever
// CLEAN provided. If a path is absent from previous (user deleted it), leaves CLEAN's
// version in place. Creates parent directories as needed for paths not present in CLEAN.
fn overlay_persistent(start: &Instant, previous_root: &str, current_root: &str, persistent_paths: &[String]) {
	println!("{} Overlaying {} persistent paths from previous state", elapsed(start), persistent_paths.len());
	for path in persistent_paths {
		let src = format!("{previous_root}{path}");
		let dst = format!("{current_root}{path}");

		// Path absent from previous state: leave whatever CLEAN provided (or nothing).
		// Use symlink_metadata so dangling symlinks are treated as present.
		if fs::symlink_metadata(&src).is_err() { continue; }

		// Parent may not exist in CLEAN (e.g. ~/.cert/ never existed in the clean image).
		if let Some(parent) = Path::new(&dst).parent() {
			if !parent.exists() {
				if let Err(e) = fs::create_dir_all(parent) {
					die(start, &format!("Failed to create parent for {dst}: {e}"));
				}
			}
		}

		// Remove CLEAN's version so cp -a can place the PREVIOUS version at the exact path.
		// (cp -a src dst with dst existing as a dir would copy src *into* dst — wrong.)
		match fs::symlink_metadata(&dst) {
			Ok(meta) if meta.is_dir() => { fs::remove_dir_all(&dst).ok(); }
			Ok(_) => { fs::remove_file(&dst).ok(); }
			Err(_) => {}
		}

		require(start, &["cp", "--reflink=always", "-a", &src, &dst]);
	}
}

fn reset(start: &Instant, name: &str, filter_path: &str, snapshots_name: &str, clean_name: &str) {
	let volume      = format!("{MOUNT_PATH}/{name}");
	let snapshots   = format!("{MOUNT_PATH}/{snapshots_name}/{name}");
	let clean       = format!("{snapshots}/{clean_name}");
	let previous    = format!("{snapshots}/PREVIOUS");
	let penultimate = format!("{snapshots}/PENULTIMATE");
	let current     = format!("{snapshots}/CURRENT");

	if Path::new(&current).is_dir() && !Path::new(&format!("{current}/.boot-ready")).is_file() {
		eprintln!("{} WRN Incomplete boot (missing .boot-ready)", elapsed(start));
		delete_subvolume(start, &current);
	}
	if !Path::new(&clean).is_dir() { die(start, &format!("CLEAN missing: {clean}")); }

	// Rotate bookkeeping snapshots: save the live volume as PREVIOUS before reset.
	if !Path::new(&penultimate).is_dir() { create_snapshot(start, &clean, &penultimate); }
	if !Path::new(&previous).is_dir() { create_snapshot(start, &clean, &previous); }
	create_snapshot(start, &previous, &penultimate);
	create_snapshot(start, &volume, &previous);

	// Build CURRENT from CLEAN (instant COW), then overlay persistent paths from PREVIOUS.
	// Original approach snapshotted PREVIOUS→CURRENT then walked the entire tree to reset
	// non-persistent entries to CLEAN. This approach avoids touching non-persistent entries
	// entirely: they come from CLEAN for free via the snapshot.
	create_snapshot(start, &clean, &current);
	require(start, &["btrfs", "property", "set", "-ts", &current, "ro", "false"]);

	let persistent_paths = parse_persistent_paths(start, filter_path);
	overlay_persistent(start, &previous, &current, &persistent_paths);

	if let Err(e) = fs::write(format!("{current}/.boot-ready"), "") {
		eprintln!("{} WRN Failed to write .boot-ready: {e}", elapsed(start));
	}

	create_snapshot(start, &current, &volume);
}

fn snapshot_only(start: &Instant, name: &str, snapshots_name: &str, clean_name: &str) {
	let volume      = format!("{MOUNT_PATH}/{name}");
	let snapshots   = format!("{MOUNT_PATH}/{snapshots_name}/{name}");
	let clean       = format!("{snapshots}/{clean_name}");
	let previous    = format!("{snapshots}/PREVIOUS");
	let penultimate = format!("{snapshots}/PENULTIMATE");
	if !Path::new(&clean).is_dir() { die(start, &format!("CLEAN missing: {clean}")); }
	if !Path::new(&penultimate).is_dir() { create_snapshot(start, &clean, &penultimate); }
	if !Path::new(&previous).is_dir() { create_snapshot(start, &clean, &previous); }
	create_snapshot(start, &previous, &penultimate);
	create_snapshot(start, &volume, &previous);
}

fn restore(start: &Instant, name: &str, snapshots_name: &str, label: &str) {
	let volume = format!("{MOUNT_PATH}/{name}");
	let source = format!("{MOUNT_PATH}/{snapshots_name}/{name}/{label}");
	if !Path::new(&source).is_dir() { die(start, &format!("Cannot restore: {source}")); }
	println!("{} Restoring {name} from {label}", elapsed(start));
	create_snapshot(start, &source, &volume);
}

fn main() {
	let start = Instant::now();
	let args: Vec<String> = env::args().collect();
	if args.len() < 5 {
		eprintln!("Usage: {} <device> <snapshots_name> <clean_name> <mode> [name=mount:filter ...]", args[0]);
		process::exit(1);
	}
	let (device, snapshots_name, clean_name, mode) = (&args[1], &args[2], &args[3], args[4].as_str());
	let pairs: Vec<(&str, &str)> = args[5..].iter().map(|arg| {
		let (name_mount, filter_path) = arg.rsplit_once(':').unwrap_or((arg, ""));
		(name_mount.split('=').next().unwrap(), filter_path)
	}).collect();

	println!("{} Mode={mode} device={device} subvolumes={}",
		elapsed(&start), pairs.iter().map(|pair| pair.0).collect::<Vec<_>>().join(" "));
	require(&start, &["mkdir", "-p", MOUNT_PATH]);
	require(&start, &["mount", "-t", "btrfs", "-o", "subvolid=5,user_subvol_rm_allowed", device, MOUNT_PATH]);

	// Guard umounts MOUNT_PATH on drop — covers both normal return and die() panic paths.
	let _unmount = Unmounter(&start);

	if mode == "disabled" {
		println!("{} Immutability disabled; skipping", elapsed(&start));
	} else {
		let dispatch = |name: &str, filter_path: &str| match mode {
			"reset"               => { println!("{} Resetting {name}", elapsed(&start)); reset(&start, name, filter_path, snapshots_name, clean_name); }
			"snapshot-only"       => { println!("{} Snapshot-only {name}", elapsed(&start)); snapshot_only(&start, name, snapshots_name, clean_name); }
			"restore-previous"    => restore(&start, name, snapshots_name, "PREVIOUS"),
			"restore-penultimate" => restore(&start, name, snapshots_name, "PENULTIMATE"),
			_ => die(&start, &format!("Unknown mode: {mode}")),
		};
		if pairs.len() > 1 {
			let owned: Vec<(String, String)> = pairs.iter().map(|(name, filter)| (name.to_string(), filter.to_string())).collect();
			// thread::scope waits for all threads before propagating any panic, so a die()
			// in one thread gives the sibling time to finish before the process exits.
			thread::scope(|scope| { for (name, filter) in &owned { scope.spawn(|| dispatch(name, filter)); } });
		} else {
			for (name, filter_path) in &pairs { dispatch(name, filter_path); }
		}
		require(&start, &["btrfs", "filesystem", "sync", MOUNT_PATH]);
	}

	println!("{} Immutability complete", elapsed(&start));
	// _unmount drops here → umount
}
